### ISSUE TITLE
How to correctly interpret MusOOEvaluator results?

### ISSUE THREAD

**Issue by calebkoy** [jpauwels/MusOOEvaluator#4]
Hello.

I'm trying to understand how to interpret the output file from a run of MusOOEvaluator. The reference file is [ground_truth.txt](https://github.com/jpauwels/MusOOEvaluator/files/1778917/ground_truth.txt), the test file is [predictions.txt](https://github.com/jpauwels/MusOOEvaluator/files/1778918/predictions.txt) and the output file is [output.txt](https://github.com/jpauwels/MusOOEvaluator/files/1778921/output.txt). I ran the following code to produce the output: `MusOOEvaluator.exe --reffile "ground_truth.txt" --testfile "predictions.txt" --chords MirexSeventhsBass --output "output.txt"`.

1. My first issue is that I don't fully understand how the value of 100.532 s has been calculated for the 'Duration of evaluated chords' (line 6 of output.txt). After studying output.txt briefly, I thought that this value was calculated by summing the durations of the unique no-chords in the reference and test files. But when I calculated this sum myself, the value was far lower than 100.532. How exactly has 'Duration of evaluated chords' been calculated?
2. Secondly, I was initially surprised that the average score was 100% (line 7), but then looked at line 15 and gather that this means that only no-chords were evaluated. Comparing the reference and text files, I see that every no-chord in the test file is also in the reference file; however, some no-chords in the reference file (lines 59, 143, 261, 332, 364, 575, 598) don't appear in the test file. Assuming my interpretation of the 100% value is correct (if not, please correct me!), why are only no-chords evaluated?
3. Finally, to further explore what was happening, I created a dummy test file, [dummy_predictions.txt](https://github.com/jpauwels/MusOOEvaluator/files/1778929/dummy_predictions.txt), which is an exact duplicate of the reference file. Evaluating the reference file against this dummy predictions file (using a similar line of code as above) gave me exactly the same output as in output.txt! Why is this? Is there something I am perhaps doing incorrectly or perhaps not understanding about how MusOOEvaluator evaluates files? I would have thought that, in this latter evaluation, the 'Correct chords' value (line 11 of output.txt), for example, should have been 100%. But it was still 0%.
4. Fourthly, the [MIREX 2017 ACE web page](url) states that WCSR is used when computing the average for a given corpus. However, in the evaluations mentioned above, the samples were randomly chosen from a data set of songs. So am I right in assuming that 'length of track' has no meaning in this case, since consecutive samples might be from completely different songs and since all the samples in a given song may not have ended up in the test file? If I am correct, then is the evaluator calculating WCSR or CSR in the evaluations I mentioned above?

I'm working to a tight university deadline, but also appreciate that you may be busy. A response as soon as you are able would be much appreciated. 

Thank you. 

Caleb

**Comment by jpauwels** [comment]
First of all, running `MusOOEvaluator.exe --reffile "ground_truth.txt" --testfile "predictions.txt" --chords MirexSeventhsBass --output "output.txt" --verbose` will produce a file `predictions.csv` that explains where the 100.532s comes from and gives you a step-by-step breakdown of the calculation.

The issue is that your start and end times are not monotonously increasing. Even when I order the chord segments both in `ground_truth.txt` and `predictions.txt` manually before running the program, the chord segments are overlapping, so the input is nonsensical, which results in nonsense output ;-)

**Comment by calebkoy** [comment]
Thank you for the clarification! I'll run the command with `--verbose`. But I still don't fully understand if I can use MusOOEvaluator as I initially intended. This is what I'm trying to do:

I have a data set of pairs `D = {(x, y)}`, where `x` is a vector corresponding to a chord segment in some song and `y` is the corresponding ground truth chord label for `x`. I then did the following:
1.  _**Randomly**_ split `D` into a training set `T` (80%) and a validation set `V` (20%).
2.  Implemented and trained a neural network using `T`. 
3. Used the trained neural net to predict the chord labels for the data items in `V`. 

You state that the start and end times should be monotonously increasing, but this isn't possible when I construct `V` by taking a random 20% of `D`! (Note: I only want to evaluate how well my neural net classifies chord labels and so I'm currently assuming that the start and end times for each `x_i` in `V` are exactly equal to the ground truth start and end times for that `x_i`.)

So the questions I have now are: 
1. Is there any way at all that I can use the MusOOEvaluator to evaluate how well my neural network has classified the chord labels for a **_randomly selected_** validation set? 
2. Or, is it the case that, to be able to use MusOOEvaluator to obtain sensible results, the chords in the validation set absolutely must have monotonously increasing start and end times? 
3. And, if 2 **_is_** the case, then am I right in assuming that the validation set `V` needs to consist of `x_i` which represent **_contiguous_** chord segments in some song (or corpus of songs)?

I really do appreciate your help with this and look forward to hearing from you soon!

**Comment by calebkoy** [comment]
I have cleared up the misunderstanding. I discussed my issue with Junqi Deng, whose work I'm trying to build on, and I'm going to re-run my experiments using a validation set `V` that contains a contiguous set of chord segments. 

Thank you again for your help, @jpauwels.

**Comment by jpauwels** [comment]
Although the evaluator is originally made for transcriptions of music pieces, it could be used for your use-case as well, if you take some considerations in account.

Currently you cut up some chord segments from different files (different timelines) and group them into one file, mixing their time stamps. This causes overlapping segments, and only one chord can be annotated (or played) at the same time. You should either use multiple reference/test files, mirroring the sources before the cut-up such that the timelines don't get mixed up, or keep using a single file but rewrite the time stamps such that they become monotonous (even better contiguous, but not absolutely necessary, see below). The absolute value of the time stamps are not really important anyway, just the chord duration, so you can simply add the duration to the end time of the previous chord. Ideally, you'd do both, so multiple reference/test files with rewritten time stamps.

Contiguity is not strictly necessary, but in case there are "gaps" (unlabelled segments), the evaluator will assume they are "silence/no-chords" (both in ground truth and prediction) and since no-chord detection also gets evaluated in the MirexSeventhsBass preset (other presets exist where this is not the case), it will influence the overall score. You can always ignore the overall score and only look at the breakdown per chord type, but using non-contiguous segments requires a bit more understanding of what's happening, so I suggest you keep things simple and rewrite your time scale(s) to make segments contiguous (see above).

The above procedure will make a random training/validation split technically possible, but from a experiment-design point of view, I would not split entirely random, but make chords from the same song (or even album, depending on your data) fall fully into either the training or validation set. This is know as grouped sampling (read http://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-for-grouped-data even if you don't use that library, couldn't immediately find a better reference). Otherwise you might get too optimistic generalisation results because if you train on a chord in a particular song, the same chord coming from the same song (played at a different time, e.g. a repetition of the chorus)  in the validation set will likely be more similar than the same chord coming from a different song. I hope this brief explanation makes sense :-)

I'll make sure to add a test for overlapping time-segments in the future, such that an error message gets returned instead of quasi-random output. This will help make the program more robust. It's always interesting to see other people come up with scenarios I didn't think of initially, so thanks for reporting!

**Comment by jpauwels** [comment]
And here's a mention @CKKOY, cause I'm not sure if you get an alert for closed issues too :-)

**Comment by calebkoy** [comment]
Thank you very much for the thorough reply! I'm going to first try your latter suggestions of grouped sampling, evaluate and analyse the results using the evaluator, and then go from there! 

And you're welcome! I'm really enjoying working on this project and am grateful that your tool exists, @jpauwels. :)
